LmDataset.py

```diff
@@ -1378,10 +1378,21 @@ def init_seq_order(self, epoch=None, seq_list=None):
     else:
       num_seqs = self._get_data_len()
       self._seq_order = self.get_seq_order_for_epoch(
-        epoch=epoch, num_seqs=num_seqs, get_seq_len=lambda i: len(self._get_data(key=self.main_source_data_key, line_nr=i)))
+        epoch=epoch, num_seqs=num_seqs,
+        get_seq_len=lambda i: self.get_estimated_seq_length(i)[self.main_source_data_key])
     self._num_seqs = len(self._seq_order)
     return True
 
+  def get_estimated_seq_length(self, seq_idx):
+    """
+    :param seq_idx: int
+    :rtype: NumbersDict
+    """
+    corpus_seq_idx = self.get_corpus_seq_idx(seq_idx)
+    assert corpus_seq_idx is not None
+
+    return NumbersDict({key: len(self._get_data(key=key, line_nr=corpus_seq_idx)) for key in self._data})
```

Comment:
Note: This will be slightly slower as before, as you load all the keys now. (But it's your dataset, so I don't care too much. If you think this does not matter - then just leave it that way.)

Also note: This will in fact block once the data is available. See the implementation of `_get_data`. `TranslationDataset` is anyway maybe an exception because it **always** loads all data into memory (this is exactly what we wanted to avoid for other datasets; all the problems we discussed are about that).

Btw, maybe you want to reimplement `TranslationDataset` later that it does not keeps everything in memory? E.g. you could go through the whole file and only store the offsets for each line (which takes much less memory, and should be much faster). And then `_load_seqs` (or directly `_get_data`) would load only the relevant parts. And the length can be estimated by the diff of the offset.

Reply:
Hmm, that's true that it is slightly slower... Maybe we could make a list of data keys an optional argument. If it is given, only load those. If not, load all that are available. But I'm also not sure if it matters.

I'm aware of the other issues. For that reason we have to use HDFDataset for large corpora.

Reply:
I wanted to keep the possibility to sort by target length, I guess. And keep it similar to `get_sequence_length()`. But actually I'm not sure whether it will be useful.

Reply:
Maybe you could also make it an argument to `get_estimated_seq_length`. Although maybe the API should define that this argument can also be ignored. So maybe an argument like `suggested_data_key`.